rfc/accepted/0021-command-driven-connectors.md

```diff
@@ -0,0 +1,119 @@
+# Command driven connectors
+
+- Feature Name: `command_driven_connectors`
+- Start Date: 2021-10-07
+- Tremor Issue: [tremor-rs/tremor-runtime#0000](https://github.com/tremor-rs/tremor-runtime/issues/0000)
+- RFC PR: [tremor-rs/tremor-rfcs#0000](https://github.com/tremor-rs/tremor-rfcs/pull/0000)
+
+## Summary
+[summary]: #summary
+
+This RFC introduces a control plane for connectors to allow controlling their behavior and actions via events.
+
+## Motivation
+[motivation]: #motivation
+
+There are many connectors, where, especially on the read side, there's no obvious "default" way to read data. For example object stores usually do not provide an API to stream changes (and if they did, the users might still want to read files in a different fashion). With command driven connectors, users will be able to read data that they need based on the commands they send to the connector.
+
+## Guide-level Explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Command-driven connectors define at least two ports - one for data and one for commands.
+As an example, let's look at a connector that reads files:
+
+```tremor
+define flow main
+flow
+  define connector file_connector from file
+    with 
+      codec="string",
+      config = {"command_driven": true}

```

Comment:
Do we really want this configurable, it means more learning surface for users and we'll have to switch "is this command going to be ignored" in every connector.

If we do this should live outside of config and at the same level as codec/config, but I think we should just not make this configurable.

Reply:
One thing that's worth considering here (not sure if it matters in practice), is that we can't synthesize events easily, so if you want a connector to read exactly one file exactly once, with command-driven connector this will require some magic (like sending the event that will generate the command from an external source).

Reply:
That's a good point, we do have `metronome`, perhaps we should add a `oneshot` or `startup` connector that send one event at the beginning of the startup sequence. It's not the first time we run into this limitation, that could be a conclusion from the RFC and it seems like a good secondary effect.